lsm-perf.py

```diff
@@ -67,68 +73,93 @@ def evaluate_kernel(kernel_path, filesystem_path, workload_path, keyfile):
+
+        :param kernel_path: Path of the kernel's bzImage
+        :param filesystem_img_path: Path of the filesystem image (.img)
+        :param keyfile: Path of rsa key that is authorized on the image
+        """
+        qemu_args = VM.__construct_qemu_args(kernel_path, filesystem_img_path)
         self.process = local['qemu-system-x86_64'].popen(qemu_args)
         self.ssh = None
         self.key = keyfile
 
     def __enter__(self):
-        """Initialize ssh connection"""
-        c = 5
+        """Initialize the ssh connection (blocks until success)"""
+        retries = 5
         while self.ssh is None:
             time.sleep(1)
             try:
                 self.ssh = SshMachine(
-                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
+                    '127.0.0.1', user='root', port=HOST_PORT, keyfile=self.key)
             except (EOFError, plumbum.machines.session.SSHCommsError) as e:
-                c -= 1
-                if c == 0:
+                retries -= 1
+                if retries == 0:
                     raise e
         return self
 
     def __exit__(self, type, value, traceback):
-        """Stops the SSH connection and the VM"""
+        """Stop the SSH connection and the VM"""
         if self.ssh is not None:
             self.ssh.close()
             self.ssh = None
-        self.process.stdin.write(QEMU_EXIT_CMD)
         self.process.terminate()
-        time.sleep(0.5)
 
     def scp_to(self, src_local, dst_remote):
-        """Send a file from the host to the VM"""
-        assert self.ssh is not None
-        fro = local.path(src_local)
-        to = self.ssh.path(dst_remote)
-        plumbum.path.utils.copy(fro, to)
-
-
-def construct_qemu_args(kernel_path, filesystem_path):
-    """Qemu arguments similar to what `vm start` produces"""
-    return [
-        '-nographic',
-        '-s',
-        '-machine', 'accel=kvm',
-        '-cpu', 'host',
-        '-device', 'e1000,netdev=net0',
-        '-netdev', 'user,id=net0,hostfwd=tcp::5555-:22',
-        '-append', 'console=ttyS0,115200 root=/dev/sda rw nokaslr',
-        '-smp', '4',
-        '-m', '4G',
-        '-drive', 'if=none,id=hd,file=%s,format=raw' % filesystem_path,
-        '-device', 'virtio-scsi-pci,id=scsi',
-        '-device', 'scsi-hd,drive=hd',
-        '-device', 'virtio-rng-pci,max-bytes=1024,period=1000',
-        '-qmp', 'tcp:localhost:4444,server,nowait',
-        '-serial', 'mon:stdio',
-        '-kernel', '%s' % kernel_path,
-        '-name', 'lsm_perf_vm,debug-threads=on'
-    ]
+        """Send a file from the host to the VM
+
+        :param src_local: local path of the file to send
+        :param dst_remote: destination path on the vm
+        :raises ValueError: when the ssh connection is not established,
+                            i.e. when not used inside a `with` block
+        """
+        if self.ssh is None:
+            raise ValueError('`VM.scp_to` must be used with an established '
```

Comment:
ValueError is widely misused in Python - according to the [docs](https://docs.python.org/3/library/exceptions.html#ValueError) it should be "Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError"

I think in this case we should define a custom error type:

```
class VMException(Exception):
    pass
```

and use that

Reply:
Ok done

Reply:
It's rather strange to have nested classes like that - for exceptions like this I would normally just rely on the name to express the fact that the exception class and the VM class are related

Reply:
Ok good to know, I updated it